File: programming 1/lab 6/src/main/java/server/managers/CommandManager.java
package server.managers;

import common.communication.*;
import server.commands.*;
import server.network.Server;
import server.commands.Command;
import server.utils.HistoryWriter;

import java.io.BufferedReader;
import java.io.IOException;
import java.time.DateTimeException;
import java.util.*;

/**
 * Менеджер, отвечающий за обработку команд и их выполнение.
 */
public class CommandManager {

    final private Map<String, Command> commands;
    final private Server server;
    private HistoryWriter historyWriter;

    public CommandManager(CollectionManager cm, Server server) {
        this.server = server;
        historyWriter = new HistoryWriter(server);

        commands = new HashMap<>();
        commands.put("help", new HelpCommand(commands));
        commands.put("add", new AddCommand(cm));
        commands.put("show", new ShowCommand(cm));
        commands.put("remove_by_id", new RemoveCommand(cm));
        commands.put("execute_script", new ExecuteScriptCommand(this));
        commands.put("exit", new ExitCommand(cm));
        commands.put("history", new HistoryCommand());
        commands.put("clear",  new ClearCommand(cm));
        commands.put("count_less_than_students_count", new CountLessCommand(cm));
        commands.put("head", new HeadCommand(cm));
        commands.put("info", new InfoCommand(cm));
        commands.put("print_descending", new PrintDescendingCommand(cm));
        commands.put("print_field_ascending_form_of_education", new PrintFieldAscendingCommand(cm));
        commands.put("remove_greater", new RemoveGreaterCommand(cm));
        commands.put("update", new UpdateCommand(cm));
    }

    /**
     * Выполняет указанную команду.
     */
    public Response execute(RequestCommand requestCommand) {
        Command command = commands.get(requestCommand.name());
        if (command != null) {
            String arg = requestCommand.arg();
            if (arg == null) arg = "";
            historyWriter.write(requestCommand.name() + " " + arg);
            try {
                return command.execute(requestCommand.arg());
            } catch (IllegalArgumentException e) {
                return new Response(ResponseStatus.ERROR, e.getMessage());
            }
        }
        else {
            return new Response(ResponseStatus.ERROR, "Команда не найдена! Введите 'help' для получения списка команд.\n");
        }
    }

    /**
     * Выполняет команду, требующую чтение данных элемента из файла.
     * @param line строка с командой и ее аргументами.
     * @param reader для чтения данных из файла.
     */
    public void execute(String line, BufferedReader reader) {
        String[] tokens = line.split(" ");
        Command command = commands.get(tokens[0]);
        if (command != null) {
            String arg = null;
            if (tokens.length > 1) arg = tokens[1];
            if (command instanceof Promptable) {
                try {
                    server.sendResponse(((Promptable) command).execute(reader, arg));
                } catch (IOException e) {
                    server.sendResponse(new Response(ResponseStatus.INTERMEDIATE_ERROR, "Ошибка ввода-вывода: " + e.getMessage()));
                } catch (NullPointerException | exceptions.EnumConstantNotFoundException | DateTimeException |
                         IllegalArgumentException e) {
                    server.sendResponse(new Response(ResponseStatus.INTERMEDIATE_ERROR, e.getMessage()));
                }
            }
            else {
                try {
                    Response response = command.execute(arg);
                    response.setStatus(ResponseStatus.INTERMEDIATE_SUCCESS);
                    server.sendResponse(response);
                } catch (IllegalArgumentException e) {
                    server.sendResponse(new Response(ResponseStatus.INTERMEDIATE_ERROR, e.getMessage()));
                }
            }
        }
        else server.sendResponse(new Response(ResponseStatus.INTERMEDIATE_ERROR, "Команда не найдена! Введите 'help' для получения списка команд.\n"));
    }
}

File: programming 1/lab 7/src/main/java/server/managers/CommandManager.kt
package server.managers

import common.communication.RequestCommand
import common.communication.Response
import common.communication.ResponseStatus
import client.exceptions.EnumConstantNotFoundException
import server.commands.*
import server.network.Server
import server.utils.HistoryWriter
import java.io.BufferedReader
import java.io.IOException
import java.time.DateTimeException

/**
 * Менеджер, отвечающий за обработку команд и их выполнение.
 */
class CommandManager(cm: CollectionManager, private val threadPoolManager: ThreadPoolManager) {
    private val commands: MutableMap<String, Command> = HashMap()

    init {
        commands["help"] = HelpCommand(commands)
        commands["add"] = AddCommand(cm)
        commands["show"] = ShowCommand(cm)
        commands["remove_by_id"] = RemoveCommand(cm)
        commands["execute_script"] = ExecuteScriptCommand(this)
        commands["exit"] = ExitCommand(cm)
        commands["history"] = HistoryCommand()
        commands["clear"] = ClearCommand(cm)
        commands["count_less_than_students_count"] = CountLessCommand(cm)
        commands["head"] = HeadCommand(cm)
        commands["info"] = InfoCommand(cm)
        commands["print_descending"] = PrintDescendingCommand(cm)
        commands["print_field_ascending_form_of_education"] = PrintFieldAscendingCommand(cm)
        commands["remove_greater"] = RemoveGreaterCommand(cm)
        commands["update"] = UpdateCommand(cm)
    }

    /**
     * Выполняет указанную команду.
     */
    fun execute(requestCommand: RequestCommand): Response {
        val command = commands[requestCommand.name]
        if (command != null) {
            val arg = requestCommand.arg
            if (arg == null) HistoryWriter.write(requestCommand.name)
            else HistoryWriter.write(requestCommand.name + " " + arg)
            return try {
                command.execute(requestCommand.arg)
            } catch (e: IllegalArgumentException) {
                Response(ResponseStatus.ERROR, e.message!!)
            }
        } else {
            return Response(ResponseStatus.ERROR, "Команда не найдена! Введите 'help' для получения списка команд.\n")
        }
    }

    /**
     * Выполняет команду, требующую чтение данных элемента из файла.
     * @param line строка с командой и ее аргументами.
     * @param reader для чтения данных из файла.
     */
    fun execute(line: String, reader: BufferedReader) {
        val tokens = line.split(' ')
        val command = commands[tokens[0]]
        if (command != null) {
            var arg: String? = null
            if (tokens.size > 1) arg = tokens[1]
            if (command is Promptable) {
                val response =
                    try {
                        (command as Promptable).execute(reader, arg)
                    } catch (e: IOException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, "Ошибка ввода-вывода: " + e.message)
                    } catch (e: NullPointerException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, e.message!!)
                    } catch (e: EnumConstantNotFoundException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, e.message!!)
                    } catch (e: DateTimeException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, e.message!!)
                    } catch (e: IllegalArgumentException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, e.message!!)
                    }
                threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(response)
                }
            } else {
                val response =
                    try {
                        Response(ResponseStatus.INTERMEDIATE_SUCCESS, command.execute(arg).descr)
                    } catch (e: IllegalArgumentException) {
                        Response(ResponseStatus.INTERMEDIATE_ERROR, e.message!!)
                    }
                threadPoolManager.responseSenderExecutor.submit {
                    Server.sendResponse(response)
                }
            }
        } else {
            threadPoolManager.responseSenderExecutor.submit {
            Server.sendResponse(Response(
                    ResponseStatus.INTERMEDIATE_ERROR,
                    "Команда не найдена! Введите 'help' для получения списка команд.\n"))
            }
        }
    }
}


